libass/ass_fontselect.c

```diff
@@ -373,21 +373,52 @@ static void free_font_info(ASS_FontProviderMetaData *meta)
 /**
  * \brief Add a font to a font provider.
  * \param provider the font provider
- * \param meta basic metadata of the font
- * \param path path to the font file, or NULL
- * \param index face index inside the file (-1 to look up by PostScript name)
- * \param data private data for the font
- * \return success
+ * \param info The font to add to the database.
+ * \return True if the font has been successfully added to the provider. Otherwise, false.
+ * \note After calling this function, **do not** call `ass_font_provider_free_fontinfo`
+ *       on the `info` parameter, as its contents have been copied.
  */
 bool
 ass_font_provider_add_font(ASS_FontProvider *provider,
-                           ASS_FontProviderMetaData *meta, const char *path,
-                           int index, void *data)
+                           ASS_FontInfo* info)
+{
+    ASS_FontSelector *selector = provider->parent;
+
+    // check size
+    if (selector->n_font >= selector->alloc_font) {
+        size_t new_size = FFMAX(1, 2 * selector->alloc_font);
+        if (!ASS_REALLOC_ARRAY(selector->font_infos, new_size))
+            return false;
+        selector->alloc_font = new_size;
+    }
+
+    ASS_FontInfo* new_font_info = selector->font_infos + selector->n_font;
+    *new_font_info = *info;
+    selector->n_font++;
+    return true;
+}
+
+/**
+ * \brief Get the font info from a provider's font.
+ * \param provider the font provider
+ * \param meta the font metadata. See struct definition for more information.
+ * \param path absolute path to font, or NULL for memory-based fonts
+ * \param index index inside a font collection file
+ *              (-1 to look up by PostScript name)
+ * \param data private data for font callbacks
+ * \return A pointer to an ASS_FontInfo corresponding to the given parameters.
+ */
+ASS_FontInfo *
+ass_font_provider_get_font_info(ASS_FontProvider *provider,
+                                ASS_FontProviderMetaData *meta, const char *path,
+                                int index, void *data)
 {
     int i;
     ASS_FontSelector *selector = provider->parent;
-    ASS_FontInfo *info = NULL;
     ASS_FontProviderMetaData implicit_meta = {0};
+    ASS_FontInfo *info = calloc(1, sizeof(ASS_FontInfo));
```

Comment:
Can we avoid this allocation by adding an output `ASS_FontInfo *` parameter to the function and retuning a boolean success/failure?
As it stands this allocates a tiny amount of memory just to discard it almost immediately after the function returned